### ISSUE TITLE
Lint for Unnecessary Manual Bounds Checking

### ISSUE THREAD

**Issue by mmstick** [Manishearth/rust-clippy#1008]
I ran across a section of translated C code that was written as such to manually avoid accessing a value out of bounds:

```rust
let hash_value = if pos + ZOPFLI_MIN_MATCH <= array.len() {
    array[pos + ZOPFLI_MIN_MATCH - 1]
} else {
    0
};
```
However, it would be better to detect this behaviour and recommend the following:

```rust
let hash_value = array.iter().nth(pos + ZOPFLI_MIN_MATCH - 1).cloned().unwrap_or(0);
```

**Comment by mcarton** [comment]
Decide between `cloned` and `map_or` might be hard (cc #443), but otherwise :+1:

**Comment by mmstick** [comment]
I would think that `cloned` would be ideal for simply copying the value in the case of `|value| *value` whereas `map_or` would be ideal for more complicated actions where you want to do more than just copying the value, like so:

``` rust
let result_is_even = array.iter().nth(index).map_or(false, |value| *value % 2 == 0);

**Comment by mcarton** [comment]
Yes but it's hard to check whether `cloned` can be called because it has a weird bound.

**Comment by mcarton** [comment]
Oh actually this is `Option::cloned` which only needs `T: Clone`. It shouldn't be a problem to check then.

**Comment by Manishearth** [comment]
Its better to use `.get()` over iterators for arrays, no? That iterator might get optimized,  but not in debug, and get is cleaner anyway.

**Comment by mcarton** [comment]
I guess we also need a lint for `s/iter().nth/get/` then :smile:
